lib/ddtrace/context.rb

```diff
@@ -90,6 +117,67 @@ def sampled?
       end
     end
 
+    # Returns ids of all spans which can be considered as local, partial roots
+    # from a partial flush perspective. Also returns the span IDs which have
+    # been marked as non flushable, and which should be kept.
+    def partial_roots
+      return nil unless @current_span
+
+      marked_ids = Hash[([@current_span.span_id] + @current_span.parent_ids).map { |id| [id, true] }]
+      roots = []
+      @trace.each do |span|
+        # Skip if span is one of the parents of the current span.
+        next if marked_ids.key? span.span_id
+        # Skip if the span is not one of the parents of the current span,
+        # and its parent is not either. It means it just can't be a local, partial root.
+        next unless marked_ids.key? span.parent_id
+
+        roots << span.span_id
+      end
+      [roots, marked_ids]
+    end
+
+    # Return a hash containting all sub traces which are candidates for
+    # a partial flush.
+    def partial_roots_spans
+      roots, marked_ids = partial_roots()
+      return nil unless roots
+
+      return unless roots
```

Comment:
Is this a duplicate of line 144?

Reply:
Thanks for spotting this, indeed, it is.